examples/ner_new/main_train.py

```diff
@@ -0,0 +1,257 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+from typing import Iterator, Dict, List
+
+import numpy as np
+import torch
+from torch import nn
+from torch.optim import SGD
+from torch.optim.optimizer import Optimizer
+from texar.torch.data import Batch
+from tqdm import tqdm
+import yaml
+
+from examples.ner_new.ner_evaluator import CoNLLNEREvaluator
+from forte.models.ner.model_factory import BiRecurrentConvCRF
+from forte.pipeline import Pipeline
+from forte.predictor import Predictor
+from forte.data.types import DATA_INPUT, DATA_OUTPUT
+from forte.common.configuration import Config
+from forte.data.extractor import AttributeExtractor
+from forte.data.extractor.base_extractor import BaseExtractor
+from forte.data.extractor.char_extractor import CharExtractor
+from forte.data.extractor.seqtagging_extractor import BioSeqTaggingExtractor
+from forte.train_preprocessor import TrainPreprocessor
+from forte.data.readers.conll03_reader_new import CoNLL03Reader
+from ft.onto.base_ontology import Sentence, Token, EntityMention
+
+logger = logging.getLogger(__name__)
+
+logging.basicConfig(level=logging.INFO)
+
+
+def construct_word_embedding_table(embed_dict, extractor: BaseExtractor):
+    embedding_dim = list(embed_dict.values())[0].shape[-1]
+
+    scale = np.sqrt(3.0 / embedding_dim)
+    table = np.empty(
+        [extractor.size(), embedding_dim], dtype=np.float32
+    )
+    oov = 0
+    for word, index in extractor.items():
+        if word in embed_dict:
+            embedding = embed_dict[word]
+        elif word.lower() in embed_dict:
+            embedding = embed_dict[word.lower()]
+        else:
+            embedding = np.random.uniform(
+                -scale, scale, [1, embedding_dim]
+            ).astype(np.float32)
+            oov += 1
+        table[index, :] = embedding
+    return torch.from_numpy(table)
+
+
+def create_model(schemes: Dict[str, Dict[str, BaseExtractor]],
+                 config: Config):
+    text_extractor: BaseExtractor = schemes["text_tag"]["extractor"]
+    char_extractor: BaseExtractor = schemes["char_tag"]["extractor"]
+    ner_extractor: BaseExtractor = schemes["ner_tag"]["extractor"]
+
+    # embedding_dict = \
+    #     load_glove_embedding(config.config_preprocessor.embedding_path)
+    #
+    # for word in embedding_dict:
+    #     if not text_extractor.has_key(word):
+    #         text_extractor.add(word)
+
+    # TODO: temporarily make fake pretrained emb for debugging
```

Comment:
Ready to switch back to real tensors?

Reply:
Now the texar word embedder is used instead of making my own.